google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/MetricsTracer.java

```diff
@@ -0,0 +1,219 @@
+  // Attempt level metrics
+  private int attemptCount = 0;
+  private Stopwatch attemptTimer;
+  private long attemptResponseCount = 0;
+
+  MetricsTracer(
+      OperationType operationType,
+      Tagger tagger,
+      StatsRecorder stats,
+      SpanName spanName,
+      Map<TagKey, TagValue> statsAttributes) {
+    this.operationType = operationType;
+    this.tagger = tagger;
+    this.stats = stats;
+    this.parentContext = tagger.getCurrentTagContext();
+    this.spanName = spanName;
+    this.statsAttributes = statsAttributes;
+  }
+
+  @Override
+  public Scope inScope() {
+    return new Scope() {
+      @Override
+      public void close() {}
+    };
+  }
+
+  @Override
+  public void operationSucceeded() {
+    recordOperationCompletion(null);
+  }
+
+  @Override
+  public void operationCancelled() {
+    recordOperationCompletion(new CancellationException());
+  }
+
+  @Override
+  public void operationFailed(Throwable throwable) {
+    recordOperationCompletion(throwable);
+  }
+
+  private void recordOperationCompletion(@Nullable Throwable throwable) {
+    if (!opFinished.compareAndSet(false, true)) {
+      return;
+    }
+    operationTimer.stop();
+
+    long elapsed = operationTimer.elapsed(TimeUnit.MILLISECONDS);
+
+    MeasureMap measures =
+        stats
+            .newMeasureMap()
+            .put(RpcMeasureConstants.BIGTABLE_OP_LATENCY, elapsed)
+            .put(RpcMeasureConstants.BIGTABLE_OP_ATTEMPT_COUNT, attemptCount);
+
+    if (operationType == OperationType.ServerStreaming
+        && spanName.getMethodName().equals("ReadRows")) {
+      measures.put(
+          RpcMeasureConstants.BIGTABLE_READ_ROWS_FIRST_ROW_LATENCY,
+          firstResponsePerOpTimer.elapsed(TimeUnit.MILLISECONDS));
+    }
+
+    TagContextBuilder tagCtx =
+        newTagCtxBuilder()
+            .putLocal(RpcMeasureConstants.BIGTABLE_STATUS, Util.extractStatus(throwable));
+
+    measures.record(tagCtx.build());
+  }
+
+  @Override
+  public void connectionSelected(String s) {
+    // noop: cardinality for connection ids is too high to use as tags
+  }
+
+  @Override
+  public void attemptStarted(int i) {
+    attemptCount++;
+    attemptTimer = Stopwatch.createStarted();
+    attemptResponseCount = 0;
+  }
+
+  @Override
+  public void attemptSucceeded() {
+    recordAttemptCompletion(null, null);
+  }
+
+  @Override
+  public void attemptCancelled() {
+    recordAttemptCompletion(new CancellationException(), null);
+  }
+
+  @Override
+  public void attemptFailed(Throwable throwable, Duration duration) {
+    recordAttemptCompletion(throwable, duration);
+  }
+
+  @Override
+  public void attemptFailedRetriesExhausted(Throwable throwable) {
+    recordAttemptCompletion(throwable, null);
+  }
+
+  @Override
+  public void attemptPermanentFailure(Throwable throwable) {
+    recordAttemptCompletion(throwable, null);
+  }
+
+  private void recordAttemptCompletion(
+      @Nullable Throwable throwable, @Nullable Duration nextAttemptDelay) {
```

Comment:
is `nextAttemptDelay` used anywhere?

Reply:
No, but it's part of the ApiTracer interface and  can be used to record delay

Reply:
Actually you are right, its unnecessary